GUEST BOOKING DETAILS

In [1]:
import requests
import json

In [ ]:
# url = f"https://restful-booker.herokuapp.com/booking/{str(i)}"
# For 1 to 10
def get_guest_details(starting_index, closing_index):
    ans = []
    for i in range(starting_index,closing_index):
        values = requests.get(f"https://restful-booker.herokuapp.com/booking/{i}")
        ans.append(values.json())
    return ans

details = get_guest_details(20,25)
for row in details:
    print(row)

ADD PRODUCTS

In [5]:
# API Endpoint : https://dummyjson.com/products/add
data = {
        'title': 'BMW Pencil',
        'price': 50.0,
        'brand': 'BMW',
        'category': 'Stationery'
}

def Add_data(data):
    string_data = json.dumps(data)
    # Get header from the given url
    header =  { 'Content-Type': 'application/json' }
    return requests.post('https://dummyjson.com/products/add',headers=header, data = string_data)

ans = Add_data(data)
print(ans.json())

{'id': 101, 'title': 'BMW Pencil', 'price': 50, 'brand': 'BMW', 'category': 'Stationery'}


#### UPDATING BOOKING DATES

In [6]:
import requests
import json

headers = {
    'Content-Type': 'application/json',
}
data = '{ "username" : "admin", "password" : "password123" }'
auth_request = requests.post('https://restful-booker.herokuapp.com/auth', headers=headers, data=data)
auth_request.status_code
x = auth_request.json()

# Before updating the booking lets first see what was the booking infomation
previous_booking_info = requests.get("https://restful-booker.herokuapp.com/booking/1")

# Here we pass the token we had generated above, please replace the token from the first request
cookies = {
    'token': x.get('token'),
}
# Your code goes here
header = {
    'Content-Type' : 'application/json',
    'Accept' : 'application/json',
    'Authorization': 'Basic YWRtaW46cGFzc3dvcmQxMjM=',
}

json_data = {'firstname': 'James',
 'lastname': 'Brown',
 'totalprice': 111,
 'depositpaid': True,
 'bookingdates': {'checkin': '2018-01-01', 'checkout': '2018-02-01'},
 'additionalneeds': 'Breakfast'}

new_data = json.dumps(json_data,indent=3)
# url = json.load("https://restful-booker.herokuapp.com/booking/1")

updated_booking_info_response = requests.put("https://restful-booker.herokuapp.com/booking/1",headers = header, data = new_data)

print(updated_booking_info_response.json())

{'firstname': 'James', 'lastname': 'Brown', 'totalprice': 111, 'depositpaid': True, 'bookingdates': {'checkin': '2018-01-01', 'checkout': '2018-02-01'}, 'additionalneeds': 'Breakfast'}


#### JSON Formatting and Parsing

In [15]:
import pandas as pd
import requests
import json

API_KEY = '7de0f426'  # Replace with your OMDB API key
BASE_URL = 'http://www.omdbapi.com/'

def fetch_movies(movie_title):
    # Your code goes here
    url = f"https://omdbapi.com/?apikey={API_KEY}&t=Harry%20Potter"
    ans = requests.get(url,)
    return ans.json()
    
    """
    Fetches movie data from OMDB API based on movie title.

    :param movie_title: Title of the movie to search.
    :return: JSON response from the API.
    """

def parse_movies(movie_data):
    # Your code goes here
    # title, release year, genre, and ratings.
    df = {}
    for i in range(len(movie_data)):
        key1 = 'Title'
        df[key1] = movie_data[key1]
        key2 = 'Year'
        df[key2] = movie_data[key2]
        key3 = 'Genre'
        df[key3] = movie_data[key3]
        key4 = 'Ratings'
        df[key4] = movie_data[key4]
    return df

    
    """
    Parses JSON movie data to extract relevant information.

    :param movie_data: JSON response from the OMDB API.
    :return: A dictionary containing relevant movie information.
    """


movie_name = "harry potter"
data = fetch_movies(movie_name)
parsed_data = parse_movies(data)
data
# data.json()['Search']
if parsed_data:
    df = pd.DataFrame([parsed_data])
    print(df)
else:
    print("Couldn't retrieve movie data.")

                                          Title  Year  \
0  Harry Potter and the Deathly Hallows: Part 2  2011   

                        Genre  \
0  Adventure, Family, Fantasy   

                                             Ratings  
0  [{'Source': 'Internet Movie Database', 'Value'...  


Problem Statement
Link: https://jsonplaceholder.typicode.com

Imagine you're building a weather application that provides real-time weather information to users. To ensure you provide the correct weather, you decide to fetch data from multiple weather APIs and then send the correct weather by combining the results. Here one might implement asynchronous programming so that instead of fetching data in a serial manner you do it parallelly 

In this exercise, we'll use asyncio to fetch multiple URLs concurrently. We'll create a list of URLs, and then use asyncio.gather() to fetch the contents of all URLs concurrently. Each URL fetching will be done asynchronously, which can result in faster execution compared to sequential fetching. Compare the time for serial and parallel execution

1. 'https://jsonplaceholder.typicode.com/posts/1',
2. 'https://jsonplaceholder.typicode.com/posts/2',
3. 'https://jsonplaceholder.typicode.com/posts/3',
4. 'https://jsonplaceholder.typicode.com/posts/4',
5. 'https://jsonplaceholder.typicode.com/posts/5',

In [4]:
import asyncio
import json
import requests

In [36]:
async def main():
    task1 = asyncio.create_task(Values("https://jsonplaceholder.typicode.com/posts/1"))
    task2 = asyncio.create_task(Values("https://jsonplaceholder.typicode.com/posts/2"))
    task3 = asyncio.create_task(Values("https://jsonplaceholder.typicode.com/posts/3"))
    task4 = asyncio.create_task(Values("https://jsonplaceholder.typicode.com/posts/4"))
    task5 = asyncio.create_task(Values("https://jsonplaceholder.typicode.com/posts/5"))

    await asyncio.gather(task1,task2,task3,task4,task5)


async def Values(url):
    print(requests.get(url).json())
    await asyncio.sleep(2)
    print("ALL")



await main()


{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}
{'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}
{'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}
{'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum culpa\nquis hic commo

* Use case: In the age of information, accessing data about different countries can be crucial for travelers, researchers, and many other professionals. While there are many APIs available to fetch this data, they come with certain limitations, like rate limits, which can restrict the number of requests one can make in a given period.
* Exercise: Your goal is to write a Python script that fetches data for a list of countries from the REST Countries API. However, you need to ensure that your script respects the rate limits of the API and can handle situations where it reaches the rate limit.
* Your script should fetch country data for the following countries: USA, Canada, Mexico, Germany, France, India, China, Japan, Australia, New Zealand, Russia, Brazil, South Africa, Egypt, Kenya, Argentina, Chile, Spain, Italy, and Portugal.
* Use the REST Countries API (https://restcountries.com/v3.1/name/) to fetch the data.
* If your script encounters a rate limit (usually indicated by a 429 HTTP status code), it should wait for a reasonable amount of time (e.g., 60 seconds) and then retry the request. Introduce a small delay between consecutive requests to minimize the risk of hitting the rate limit. If data is successfully fetched for a country, print the data.

In [2]:
import requests
import time

BASE_URL = "https://restcountries.com/v3.1/name/"  # URL endpoint for REST Countries API

def fetch_country_data(country_name):
    # Your code goes here
    ans = requests.get(BASE_URL+str(country_name))
    if ans.status_code == 200:
        return ans.json()
    elif ans.status_code == 429:
        time.sleep(60)
        ans = requests.get(BASE_URL+str(country_name))
        return ans.json()
    else:
        return None
    """
    Fetches country data from the REST Countries API.

    :param country_name: Name of the country to search.
    :return: JSON response from the API or None if any issues.
    """



countries = [
    'USA', 'Canada', 'Mexico', 'Germany', 'France', 
    'India', 'China', 'Japan', 'Australia', 'New Zealand', 
    'Russia', 'Brazil', 'South Africa', 'Egypt', 'Kenya', 
    'Argentina', 'Chile', 'Spain', 'Italy', 'Portugal'
] 

for country in countries:
    data = fetch_country_data(country)
    if data:
        print(data[0]["name"])


{'common': 'United States', 'official': 'United States of America', 'nativeName': {'eng': {'official': 'United States of America', 'common': 'United States'}}}
{'common': 'Canada', 'official': 'Canada', 'nativeName': {'eng': {'official': 'Canada', 'common': 'Canada'}, 'fra': {'official': 'Canada', 'common': 'Canada'}}}
{'common': 'Mexico', 'official': 'United Mexican States', 'nativeName': {'spa': {'official': 'Estados Unidos Mexicanos', 'common': 'México'}}}
{'common': 'Germany', 'official': 'Federal Republic of Germany', 'nativeName': {'deu': {'official': 'Bundesrepublik Deutschland', 'common': 'Deutschland'}}}
{'common': 'France', 'official': 'French Republic', 'nativeName': {'fra': {'official': 'République française', 'common': 'France'}}}
{'common': 'British Indian Ocean Territory', 'official': 'British Indian Ocean Territory', 'nativeName': {'eng': {'official': 'British Indian Ocean Territory', 'common': 'British Indian Ocean Territory'}}}
{'common': 'Macau', 'official': "Macao S

#### Problem Statement
* Imagine you're developing a BlogPost Aggregator application that gathers and organizes blog post data from multiple sources using APIs. Your goal is to fetch and display blog post information, including titles and authors, in an organized manner.

* Link: https://jsonplaceholder.typicode.com/
#### Exercise
Your application needs to perform the following tasks:

1. Fetch Users: Retrieve a list of users from the API.
2. Fetch Posts: Fetch a list of blog posts from the API.
3. Fetch Comments: For each post, fetch the associated comments.
4. Organize Data: Organize the fetched data to display blog post titles, authors, and the number of comments each post has.change column name id to userId in users,id to postId in posts and id to commentId,name to commented_by_name, email to commented_by_email, body to comment_body in comments. combine the data
5. Write a Python script that utilizes the APIs available at https://jsonplaceholder.typicode.com to achieve the tasks mentioned above.


In [19]:
import requests
link_mapping = {
    "users": "https://jsonplaceholder.typicode.com/users",
    "posts": "https://jsonplaceholder.typicode.com/posts",
    "comments": "https://jsonplaceholder.typicode.com/comments"
}

User_data = requests.get(link_mapping["users"])
Posts = requests.get(link_mapping["posts"])
Comments = requests.get(link_mapping["comments"])


In [23]:
print(User_data)

<Response [200]>


In [24]:
for i in User_data.json():
    ans = f"{str(i['id'])}"
    # ans = f
    print(type(ans))
    # print(str(i['id']))

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


* comments_data = 'postId', 'id', 'name', 'email', 'body'
* Posts_data = 'userId','id','title','body'
* User_daat = id,name,username,email,
    address: {'street','suite','city','zipcode','geo','lng'},
    'phone','website',
    'company': {'name','catchPhrase','bs'}



1.  userId
2.    name
3.    username
4.    email
5.    address
6.    phone
7.    website
8.    company
9.    postId
10.    title
11.    body
12.    commentId
13.    commented_by_name
14.    commented_by_email
15.    comment_body

In [25]:
userId = []
name = []
username = []
email = []
address = []
phone = []
website = []
company = []


for i in User_data.json():
    userId.append(i['id'])
    name.append(i['name'])
    username.append(i['username'])
    email.append(i['email'])
    address.append(i['address'])
    phone.append(i['phone'])
    website.append(i['website'])
    company.append(i['company'])
print(userId,name)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10] ['Leanne Graham', 'Ervin Howell', 'Clementine Bauch', 'Patricia Lebsack', 'Chelsey Dietrich', 'Mrs. Dennis Schulist', 'Kurtis Weissnat', 'Nicholas Runolfsdottir V', 'Glenna Reichert', 'Clementina DuBuque']


In [26]:
postId = []
title = []
body = []
for i in Posts.json():
    postId.append(i['id'])
    title.append(i['title'])
    body.append(i['body'])

postId = postId[:10]
title = title[:10]
body = body[:10]

In [27]:
commentId = []
commented_by_name = []
commented_by_email = []
comment_body = []
for i in Comments.json():
    commentId.append(i['id'])
    commented_by_name.append(i['name'])
    commented_by_email.append(i['email'])
    comment_body.append(i['body'])

commentId = commentId[:10]
commented_by_name = commented_by_name[:10]
commented_by_email = commented_by_email[:10]
comment_body = comment_body[:10]

# print(commentId,commented_by_name,commented_by_email,comment_body)

In [28]:
comment_body

['laudantium enim quasi est quidem magnam voluptate ipsam eos\ntempora quo necessitatibus\ndolor quam autem quasi\nreiciendis et nam sapiente accusantium',
 'est natus enim nihil est dolore omnis voluptatem numquam\net omnis occaecati quod ullam at\nvoluptatem error expedita pariatur\nnihil sint nostrum voluptatem reiciendis et',
 'quia molestiae reprehenderit quasi aspernatur\naut expedita occaecati aliquam eveniet laudantium\nomnis quibusdam delectus saepe quia accusamus maiores nam est\ncum et ducimus et vero voluptates excepturi deleniti ratione',
 'non et atque\noccaecati deserunt quas accusantium unde odit nobis qui voluptatem\nquia voluptas consequuntur itaque dolor\net qui rerum deleniti ut occaecati',
 'harum non quasi et ratione\ntempore iure ex voluptates in ratione\nharum architecto fugit inventore cupiditate\nvoluptates magni quo et',
 'doloribus at sed quis culpa deserunt consectetur qui praesentium\naccusamus fugiat dicta\nvoluptatem rerum ut voluptate autem\nvoluptatem 

In [29]:
import pandas as pd

In [30]:
dict = {'userId' : userId,'name' : name,'username' : username,
        'email' : email,'address' : address,'phone' : phone,
        'website' : website,'company' : company,'postId' : postId,
        'title' : title,'body' : body,'commentId' : commentId,
        'commented_by_name' : commented_by_name,'commented_by_email' : commented_by_email,
        'comment_body' : comment_body
}

In [31]:
df = pd.DataFrame(dict)

In [32]:
df

userId                      name          username  \
0       1             Leanne Graham              Bret   
1       2              Ervin Howell         Antonette   
2       3          Clementine Bauch          Samantha   
3       4          Patricia Lebsack          Karianne   
4       5          Chelsey Dietrich            Kamren   
5       6      Mrs. Dennis Schulist  Leopoldo_Corkery   
6       7           Kurtis Weissnat      Elwyn.Skiles   
7       8  Nicholas Runolfsdottir V     Maxime_Nienow   
8       9           Glenna Reichert          Delphine   
9      10        Clementina DuBuque    Moriah.Stanton   

                       email  \
0          Sincere@april.biz   
1          Shanna@melissa.tv   
2         Nathan@yesenia.net   
3  Julianne.OConner@kory.org   
4   Lucio_Hettinger@annie.ca   
5    Karley_Dach@jasper.info   
6     Telly.Hoeger@billy.biz   
7       Sherwood@rosamond.me   
8    Chaim_McDermott@dana.io   
9     Rey.Padberg@karina.biz   

                                             address                  phone  \
0  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
1  {'street': 'Victor Plains', 'suite': 'Suite 87...    010-692-6593 x09125   
2  {'street': 'Douglas Extension', 'suite': 'Suit...         1-463-123-4447   
3  {'street': 'Hoeger Mall', 'suite': 'Apt. 692',...      493-170-9623 x156   
4  {'street': 'Skiles Walks', 'suite': 'Suite 351...          (254)954-1289   
5  {'street': 'Norberto Crossing', 'suite': 'Apt....   1-477-935-8478 x6430   
6  {'street': 'Rex Trail', 'suite': 'Suite 280', ...           210.067.6132   
7  {'street': 'Ellsworth Summit', 'suite': 'Suite...      586.493.6943 x140   
8  {'street': 'Dayna Park', 'suite': 'Suite 449',...   (775)976-6794 x41206   
9  {'street': 'Kattie Turnpike', 'suite': 'Suite ...           024-648-3804   

         website                                            company  postId  \
0  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
1  anastasia.net  {'name': 'Deckow-Crist', 'catchPhrase': 'Proac...       2   
2    ramiro.info  {'name': 'Romaguera-Jacobson', 'catchPhrase': ...       3   
3       kale.biz  {'name': 'Robel-Corkery', 'catchPhrase': 'Mult...       4   
4   demarco.info  {'name': 'Keebler LLC', 'catchPhrase': 'User-c...       5   
5        ola.org  {'name': 'Considine-Lockman', 'catchPhrase': '...       6   
6       elvis.io  {'name': 'Johns Group', 'catchPhrase': 'Config...       7   
7   jacynthe.com  {'name': 'Abernathy Group', 'catchPhrase': 'Im...       8   
8     conrad.com  {'name': 'Yost and Sons', 'catchPhrase': 'Swit...       9   
9    ambrose.net  {'name': 'Hoeger LLC', 'catchPhrase': 'Central...      10   

                                               title  \
0  sunt aut facere repellat provident occaecati e...   
1                                       qui est esse   
2  ea molestias quasi exercitationem repellat qui...   
3                               eum et est occaecati   
4                                 nesciunt quas odio   
5                 dolorem eum magni eos aperiam quia   
6                               magnam facilis autem   
7                           dolorem dolore est ipsam   
8  nesciunt iure omnis dolorem tempora et accusan...   
9                        optio molestias id quia eum   

                                                body  commentId  \
0  quia et suscipit\nsuscipit recusandae consequu...          1   
1  est rerum tempore vitae\nsequi sint nihil repr...          2   
2  et iusto sed quo iure\nvoluptatem occaecati om...          3   
3  ullam et saepe reiciendis voluptatem adipisci\...          4   
4  repudiandae veniam quaerat sunt sed\nalias aut...          5   
5  ut aspernatur corporis harum nihil quis provid...          6   
6  dolore placeat quibusdam ea quo vitae\nmagni q...          7   
7  dignissimos aperiam dolorem qui eum\nfacilis q...          8   
8  consectetur animi nesciunt iure dolore\nenim q...          

userId,name,username,email,address,phone,website,company,postId,title,body,commentId,commented_by_name,commented_by_email,comment_body
1,Leanne Graham,Bret,Sincere@april.biz,"{'street': 'Kulas Light', 'suite': 'Apt. 556', 'city': 'Gwenborough', 'zipcode': '92998-3874', 'geo': {'lat': '-37.3159', 'lng': '81.1496'}}",1-770-736-8031 x56442,hildegard.org,"{'name': 'Romaguera-Crona', 'catchPhrase': 'Multi-layered client-server neural-net', 'bs': 'harness real-time e-markets'}",1,sunt aut facere repellat provident occaecati excepturi optio reprehenderit,"quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",1,id labore ex et quam laborum,Eliseo@gardner.biz,"laudantium enim quasi est quidem magnam voluptate ipsam eos
tempora quo necessitatibus
dolor quam autem quasi
reiciendis et nam sapiente accusantium ...

In [43]:
import pandas as pd
import requests

link_mapping = {
    "users": "https://jsonplaceholder.typicode.com/users",
    "posts": "https://jsonplaceholder.typicode.com/posts",
    "comments": "https://jsonplaceholder.typicode.com/comments"
}

def fetch_link_and_return_df(link: str) -> pd.DataFrame:
    # Your code goes here
    User_data = requests.get(link)
    print(f"Get Response for link {link} is {User_data.status_code}")
    return pd.DataFrame.from_records(User_data.json())
    """
    arg:
        link: URL-Endpoint for which the data needs to be fetched from the website `https://jsonplaceholder.typicode.com`
    """
df_mapping = {}
for key in link_mapping:
    df_mapping[key] = fetch_link_and_return_df(link_mapping[key])

# Process User DataFrames
users_df = df_mapping["users"]
users_df = users_df.rename(columns={"id": "userId"})

# Process Posts DataFrames
posts_df = df_mapping["posts"]
posts_df = posts_df.rename(columns={"id": "postId"})

# Process Comments DataFrames
comments_df = df_mapping["comments"]
comments_df = comments_df.rename(columns={"id": "commentId", "name": "commented_by_name", "email": "commented_by_email", "body": "comment_body"})

# Merge the three dataframes
combined_dataframe = users_df.merge(posts_df, on=["userId"]).merge(comments_df, on=["postId"])

# Display the first few rows of the combined DataFrame
combined_dataframe.head()


Get Response for link https://jsonplaceholder.typicode.com/users is 200
Get Response for link https://jsonplaceholder.typicode.com/posts is 200
Get Response for link https://jsonplaceholder.typicode.com/comments is 200


userId           name username              email  \
0       1  Leanne Graham     Bret  Sincere@april.biz   
1       1  Leanne Graham     Bret  Sincere@april.biz   
2       1  Leanne Graham     Bret  Sincere@april.biz   
3       1  Leanne Graham     Bret  Sincere@april.biz   
4       1  Leanne Graham     Bret  Sincere@april.biz   

                                             address                  phone  \
0  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
1  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
2  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
3  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
4  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   

         website                                            company  postId  \
0  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
1  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
2  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
3  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
4  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   

                                               title  \
0  sunt aut facere repellat provident occaecati e...   
1  sunt aut facere repellat provident occaecati e...   
2  sunt aut facere repellat provident occaecati e...   
3  sunt aut facere repellat provident occaecati e...   
4  sunt aut facere repellat provident occaecati e...   

                                                body  commentId  \
0  quia et suscipit\nsuscipit recusandae consequu...          1   
1  quia et suscipit\nsuscipit recusandae consequu...          2   
2  quia et suscipit\nsuscipit recusandae consequu...          3   
3  quia et suscipit\nsuscipit recusandae consequu...          4   
4  quia et suscipit\nsuscipit recusandae consequu...          5   

                           commented_by_name      commented_by_email  \
0               id labore ex et quam laborum      Eliseo@gardner.biz   
1  quo vero reiciendis velit similique earum  Jayne_Kuhic@sydney.com   
2              odio adipisci rerum aut animi     Nikita@garfield.biz   
3                             alias odio sit           Lew@alysha.tv   
4      vero eaque aliquid doloribus et culpa       Hayden@althea.biz   

                                        comment_body  
0  laudantium enim quasi est quidem magnam volupt...  
1  est natus enim nihil est dolore omnis voluptat...  
2  quia molestiae reprehenderit quasi aspernatur\...  
3  non et atque\noccaecati deserunt quas accusant...  
4  harum non quasi et ratione\ntempore iure ex vo...

In [38]:
import pandas as pd
import requests

link_mapping = {
    "users": "https://jsonplaceholder.typicode.com/users",
    "posts": "https://jsonplaceholder.typicode.com/posts",
    "comments": "https://jsonplaceholder.typicode.com/comments"
}

def fetch_link_and_return_df(link: str) -> pd.DataFrame:
    response_from_link = requests.get(link)
    print(f"Get Response for link {link} is {response_from_link.status_code}")
    df_from_response = pd.DataFrame.from_records(response_from_link.json())
    return df_from_response

df_mapping = {}
for key in link_mapping:
    df_mapping[key] = fetch_link_and_return_df(link_mapping[key])

# Process User DataFrames
users_df = df_mapping["users"]
users_df = users_df.rename(columns={"id": "userId"})

# Process Posts DataFrames
posts_df = df_mapping["posts"]
posts_df = posts_df.rename(columns={"id": "postId"})

# Process Comments DataFrames
comments_df = df_mapping["comments"]
comments_df = comments_df.rename(columns={"id": "commentId", "name": "commented_by_name", "email": "commented_by_email", "body": "comment_body"})

# Merge the three dataframes
combined_dataframe = users_df.merge(posts_df, on=["userId"]).merge(comments_df, on=["postId"])

# Display the first few rows of the combined DataFrame
combined_dataframe.head()

Get Response for link https://jsonplaceholder.typicode.com/users is 200
Get Response for link https://jsonplaceholder.typicode.com/posts is 200
Get Response for link https://jsonplaceholder.typicode.com/comments is 200


userId           name username              email  \
0       1  Leanne Graham     Bret  Sincere@april.biz   
1       1  Leanne Graham     Bret  Sincere@april.biz   
2       1  Leanne Graham     Bret  Sincere@april.biz   
3       1  Leanne Graham     Bret  Sincere@april.biz   
4       1  Leanne Graham     Bret  Sincere@april.biz   

                                             address                  phone  \
0  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
1  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
2  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
3  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   
4  {'street': 'Kulas Light', 'suite': 'Apt. 556',...  1-770-736-8031 x56442   

         website                                            company  postId  \
0  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
1  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
2  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
3  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   
4  hildegard.org  {'name': 'Romaguera-Crona', 'catchPhrase': 'Mu...       1   

                                               title  \
0  sunt aut facere repellat provident occaecati e...   
1  sunt aut facere repellat provident occaecati e...   
2  sunt aut facere repellat provident occaecati e...   
3  sunt aut facere repellat provident occaecati e...   
4  sunt aut facere repellat provident occaecati e...   

                                                body  commentId  \
0  quia et suscipit\nsuscipit recusandae consequu...          1   
1  quia et suscipit\nsuscipit recusandae consequu...          2   
2  quia et suscipit\nsuscipit recusandae consequu...          3   
3  quia et suscipit\nsuscipit recusandae consequu...          4   
4  quia et suscipit\nsuscipit recusandae consequu...          5   

                           commented_by_name      commented_by_email  \
0               id labore ex et quam laborum      Eliseo@gardner.biz   
1  quo vero reiciendis velit similique earum  Jayne_Kuhic@sydney.com   
2              odio adipisci rerum aut animi     Nikita@garfield.biz   
3                             alias odio sit           Lew@alysha.tv   
4      vero eaque aliquid doloribus et culpa       Hayden@althea.biz   

                                        comment_body  
0  laudantium enim quasi est quidem magnam volupt...  
1  est natus enim nihil est dolore omnis voluptat...  
2  quia molestiae reprehenderit quasi aspernatur\...  
3  non et atque\noccaecati deserunt quas accusant...  
4  harum non quasi et ratione\ntempore iure ex vo...